error: use of overloaded operator '&' is ambiguous

```
module SS7LayersPcapOrNg;
function reverseOrderDigits(onePast: uint8, thisOne: uint8) {
}

function label_then_BCD(contentExceptLast: bytes, lastByte: uint8) : string {
  for (idx in contentExceptLast) {
    reverseOrderDigits(idx, uint8(0xF) & (idx >> 4));
  }
  return "%s" % (lastByte,);
}
```
reports
```
3380 /usr/local/etc/tmp$spicy/build/bin/spicyc -j Issue_219_bitwise_ambiguous.spicy
SS7LayersPcapOrNg.cc:21:78: error: use of overloaded operator '&' is ambiguous (with operand types 'hilti::rt::integer::safe<std::uint8_t>' (aka 'SafeInt<unsigned char, detail::SafeIntException>') and 'SafeInt<char, hilti::rt::integer::detail::SafeIntException>')
        reverseOrderDigits(idx, (hilti::rt::integer::safe<std::uint8_t>{15u} & idx >> hilti::rt::integer::safe<std::uint64_t>{4u}));
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/etc/tmp/spicy/hilti/include/hilti/3rdparty/SafeInt/SafeInt.hpp:6222:34: note: candidate function [with U = SafeInt<char, hilti::rt::integer::detail::SafeIntException>]
    _CONSTEXPR14 SafeInt< T, E > operator &( U rhs ) const SAFEINT_NOTHROW
                                 ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/3rdparty/SafeInt/SafeInt.hpp:7080:30: note: candidate function [with T = char, U = SafeInt<unsigned char, hilti::rt::integer::detail::SafeIntException>, E = hilti::rt::integer::detail::SafeIntException]
_CONSTEXPR11 SafeInt< T, E > operator &( U lhs, SafeInt< T, E > rhs ) SAFEINT_NOTHROW
                             ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/3rdparty/SafeInt/SafeInt.hpp:6216:34: note: candidate function
    _CONSTEXPR14 SafeInt< T, E > operator &( SafeInt< T, E > rhs ) const SAFEINT_NOTHROW
                                 ^
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, int)
        reverseOrderDigits(idx, (hilti::rt::integer::safe<std::uint8_t>{15u} & idx >> hilti::rt::integer::safe<std::uint64_t>{4u}));
                                                                             ^
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned __int128)
1 error generated.
[error] jit: failed to execute compilation action.
[error] jit: failed to compile C++ code unit SS7LayersPcapOrNg to bitcode
[error] spicyc: JIT compilation failed
```

--- a/hilti/CMakeLists.txt
+++ b/hilti/CMakeLists.txt
@@ -267,6 +267,7 @@ target_link_libraries(jit-test PRIVATE hilti)
 add_executable(hilti-rt-tests
                src/rt/tests/main.cc
                src/rt/tests/address.cc
+               src/rt/tests/bytes.cc
                src/rt/tests/fiber.cc
                src/rt/tests/reference.cc
                src/rt/tests/result.cc

--- a/hilti/include/rt/types/bytes.h
+++ b/hilti/include/rt/types/bytes.h
@@ -94,13 +94,14 @@ inline std::ostream& operator<<(std::ostream& out, const SafeConstIterator& /* x
 /**
  * HILTI's bytes instance, built on top of a std::string.
  */
-class Bytes : public std::string, public hilti::rt::detail::iterator::Controllee {
+class Bytes : protected std::string, public hilti::rt::detail::iterator::Controllee {
 public:
     using Base = std::string;
     using Iterator = bytes::SafeConstIterator;
     using Offset = uint64_t;
 
     using Base::Base;
+    using Base::data;
 
     /**
      * Creates a bytes instance from a UTF8 string, transforming the contents
@@ -129,13 +130,13 @@ public:
     /** Returns the bytes' data as a string instance. */
     const std::string& str() const { return *this; }
 
-    /** Returns an interator representing the first byte of the instance. */
+    /** Returns an iterator representing the first byte of the instance. */
     Iterator begin() const { return {*this, Base::begin()}; }
 
-    /** Returns an interator representing the end of the instance. */
+    /** Returns an iterator representing the end of the instance. */
     Iterator end() const { return Iterator(*this, Base::end()); }
 
-    /** Returns an iterator refering to the given offset. */
+    /** Returns an iterator referring to the given offset. */
     Iterator at(Offset o) const { return begin() + o; }
 
     /** Returns true if the data's size is zero. */
@@ -175,7 +176,7 @@ public:
      * @param from iterator pointing to start of subrange
      * @param to iterator pointing to just beyond subrange
      */
-    Bytes sub(const Iterator& from, const Iterator& to) const { return substr(from - begin(), to - from); }
+    Bytes sub(const Iterator& from, const Iterator& to) const { return {substr(from - begin(), to - from)}; }
 
     /**
      * Extracts a subrange of bytes from the beginning.
@@ -190,7 +191,7 @@ public:
      * @param offset of start of subrage
      * @param offset of one byeond end of subrage
      */
-    Bytes sub(Offset from, Offset to) const { return substr(from, to - from); }
+    Bytes sub(Offset from, Offset to) const { return {substr(from, to - from)}; }
 
     /**
      * Extracts a subrange of bytes from the beginning.
@@ -257,7 +258,7 @@ public:
     Vector<Bytes> split() {
         Vector<Bytes> x;
         for ( auto& v : hilti::rt::split(*this) )
-            x.push_back(Bytes(v));
+            x.emplace_back(Bytes::Base(v));
         return x;
     }
 
@@ -274,7 +275,7 @@ public:
     Vector<Bytes> split(const Bytes& sep) const {
         Vector<Bytes> x;
         for ( auto& v : hilti::rt::split(*this, sep) )
-            x.push_back(Bytes(v));
+            x.push_back(Bytes::Base(v));
         return x;
     }
 
@@ -368,6 +369,34 @@ public:
      * @return the matching group, or unset if no match
      */
     Result<Bytes> match(const RegExp& re, unsigned int group = 0);
+
+    // Add some operators over `Base`.
+    friend bool operator==(const Bytes& a, const Bytes& b) {
+        return static_cast<const Bytes::Base&>(a) == static_cast<const Bytes::Base&>(b);
+    }
+
+    friend bool operator!=(const Bytes& a, const Bytes& b) { return ! (a == b); }
+
+
+    friend bool operator<(const Bytes& a, const Bytes& b) {
+        return static_cast<const Bytes::Base&>(a) < static_cast<const Bytes::Base&>(b);
+    }
+
+    friend bool operator<=(const Bytes& a, const Bytes& b) {
+        return static_cast<const Bytes::Base&>(a) <= static_cast<const Bytes::Base&>(b);
+    }
+
+    friend bool operator>(const Bytes& a, const Bytes& b) {
+        return static_cast<const Bytes::Base&>(a) > static_cast<const Bytes::Base&>(b);
+    }
+
+    friend bool operator>=(const Bytes& a, const Bytes& b) {
+        return static_cast<const Bytes::Base&>(a) >= static_cast<const Bytes::Base&>(b);
+    }
+
+    friend Bytes operator+(const Bytes& a, const Bytes& b) {
+        return static_cast<const Bytes::Base&>(a) + static_cast<const Bytes::Base&>(b);
+    }
 };
 
 inline std::ostream& operator<<(std::ostream& out, const Bytes& x) {
@@ -377,7 +406,7 @@ inline std::ostream& operator<<(std::ostream& out, const Bytes& x) {
 
 namespace bytes {
 inline namespace literals {
-inline Bytes operator"" _b(const char* str, size_t size) { return Bytes(str, size); }
+inline Bytes operator"" _b(const char* str, size_t size) { return Bytes(Bytes::Base(str, size)); }
 } // namespace literals
 } // namespace bytes
 
@@ -387,8 +416,8 @@ inline std::string detail::to_string_for_print<Bytes>(const Bytes& x) {
 }
 
 namespace detail::adl {
-inline auto begin(const Bytes& x, adl::tag /*unused*/) { return x.begin(); }
-inline auto end(const Bytes& x, adl::tag /*unused*/) { return x.end(); }
+inline auto safe_begin(const Bytes& x, adl::tag /*unused*/) { return x.begin(); }
+inline auto safe_end(const Bytes& x, adl::tag /*unused*/) { return x.end(); }
 inline std::string to_string(const Bytes& x, adl::tag /*unused*/) { return fmt("b\"%s\"", escapeBytes(x.str(), true)); }
 } // namespace detail::adl
 

--- a/hilti/include/rt/types/bytes.h
+++ b/hilti/include/rt/types/bytes.h
@@ -130,13 +130,13 @@ public:
     const std::string& str() const { return *this; }
 
     /** Returns an interator representing the first byte of the instance. */
-    Iterator safeBegin() const { return {*this, Base::begin()}; }
+    Iterator begin() const { return {*this, Base::begin()}; }
 
     /** Returns an interator representing the end of the instance. */
-    Iterator safeEnd() const { return Iterator(*this, Base::end()); }
+    Iterator end() const { return Iterator(*this, Base::end()); }
 
     /** Returns an iterator refering to the given offset. */
-    Iterator at(Offset o) const { return safeBegin() + o; }
+    Iterator at(Offset o) const { return begin() + o; }
 
     /** Returns true if the data's size is zero. */
     bool isEmpty() const { return empty(); }
@@ -151,10 +151,10 @@ public:
      * @param n optional starting point, which must be inside the same instance
      */
     Iterator find(value_type b, const Iterator& n = Iterator()) const {
-        if ( auto i = Base::find(b, (n ? n - safeBegin() : 0)); i != Base::npos )
-            return safeBegin() + i;
+        if ( auto i = Base::find(b, (n ? n - begin() : 0)); i != Base::npos )
+            return begin() + i;
         else
-            return safeEnd();
+            return end();
     }
 
     /**
@@ -175,14 +175,14 @@ public:
      * @param from iterator pointing to start of subrange
      * @param to iterator pointing to just beyond subrange
      */
-    Bytes sub(const Iterator& from, const Iterator& to) const { return substr(from - safeBegin(), to - from); }
+    Bytes sub(const Iterator& from, const Iterator& to) const { return substr(from - begin(), to - from); }
 
     /**
      * Extracts a subrange of bytes from the beginning.
      *
      * @param to iterator pointing to just beyond subrange
      */
-    Bytes sub(const Iterator& to) const { return sub(safeBegin(), to); }
+    Bytes sub(const Iterator& to) const { return sub(begin(), to); }
 
     /**
      * Extracts a subrange of bytes.
@@ -387,8 +387,8 @@ inline std::string detail::to_string_for_print<Bytes>(const Bytes& x) {
 }
 
 namespace detail::adl {
-inline auto safe_begin(const Bytes& x, adl::tag /*unused*/) { return x.safeBegin(); }
-inline auto safe_end(const Bytes& x, adl::tag /*unused*/) { return x.safeEnd(); }
+inline auto begin(const Bytes& x, adl::tag /*unused*/) { return x.begin(); }
+inline auto end(const Bytes& x, adl::tag /*unused*/) { return x.end(); }
 inline std::string to_string(const Bytes& x, adl::tag /*unused*/) { return fmt("b\"%s\"", escapeBytes(x.str(), true)); }
 } // namespace detail::adl
 

--- a/hilti/include/rt/types/bytes.h
+++ b/hilti/include/rt/types/bytes.h
@@ -39,8 +39,10 @@ public:
     using Base = hilti::rt::detail::iterator::SafeIterator<Bytes, std::string::iterator, SafeIterator>;
     using Base::Base;
 
+    using reference = uint8_t;
+
     // Override to return expected type.
-    uint8_t operator*() const {
+    reference operator*() const {
         ensureValid();
         return *iterator();
     }
@@ -52,8 +54,10 @@ public:
     using Base = hilti::rt::detail::iterator::SafeIterator<Bytes, std::string::const_iterator, SafeConstIterator>;
     using Base::Base;
 
+    using reference = uint8_t;
+
     // Override to return expected type.
-    uint8_t operator*() const {
+    reference operator*() const {
         ensureValid();
         return *iterator();
     }

--- a/hilti/src/rt/tests/bytes.cc
+++ b/hilti/src/rt/tests/bytes.cc
@@ -1,0 +1,20 @@
+// Copyright (c) 2020 by the Zeek Project. See LICENSE for details.
+
+#include <doctest/doctest.h>
+
+#include <type_traits>
+
+#include <hilti/rt/types/bytes.h>
+
+using namespace hilti::rt;
+
+TEST_CASE("Bytes") {
+    SUBCASE("iteration") {
+        // Validate that when iterating we yield the `Iterator`'s `reference` type.
+        // This is a regression test for #219.
+        for ( auto x : Bytes() ) {
+            (void)x;
+            static_assert(std::is_same_v<decltype(x), Bytes::Iterator::reference>);
+        }
+    }
+}

--- a/hilti/src/rt/tests/stream.cc
+++ b/hilti/src/rt/tests/stream.cc
@@ -15,6 +15,8 @@ using namespace hilti::rt::stream;
 using hilti::rt::to_string;
 
 TEST_CASE("Constructors") {
+    auto b = "xyz"_b;
+    CHECK(b.size());
     auto x = Stream("xyz"_b);
     CHECK(to_string(x) == R"(b"xyz")");
     CHECK(! x.isEmpty());
@@ -197,7 +199,7 @@ TEST_CASE("sub") {
 
     auto f = [](const stream::View& v) { return v.sub(v.safeBegin() + 15, v.safeBegin() + 25); };
 
-    CHECK(f(x.view()).data() == "6789012345"_b);
+    CHECK(Bytes(f(x.view()).data()) == "6789012345"_b);
 }
 
 TEST_CASE("freezing") {
@@ -320,10 +322,10 @@ TEST_CASE("Block iteration") {
     CHECK(block->is_last);
     CHECK(! v.nextBlock(block));
 
-    x.append("567");
-    x.append("890");
-    x.append("abc");
-    x.append("def");
+    x.append("567"_b);
+    x.append("890"_b);
+    x.append("abc"_b);
+    x.append("def"_b);
 
     v = x.view();
     block = v.firstBlock();

--- a/hilti/src/rt/types/bytes.cc
+++ b/hilti/src/rt/types/bytes.cc
@@ -10,28 +10,28 @@ using namespace hilti::rt::bytes;
 
 std::tuple<bool, Bytes::Iterator> Bytes::find(const Bytes& v, const Iterator& n) const {
     if ( v.isEmpty() )
-        return std::make_tuple(true, n ? n : safeBegin());
+        return std::make_tuple(true, n ? n : begin());
 
-    auto first = *v.safeBegin();
+    auto first = *v.begin();
 
-    for ( auto i = Iterator(n ? n : safeBegin()); true; ++i ) {
-        if ( i == safeEnd() )
+    for ( auto i = Iterator(n ? n : begin()); true; ++i ) {
+        if ( i == end() )
             return std::make_tuple(false, i);
 
         if ( *i != first )
             continue;
 
         auto x = i;
-        auto y = v.safeBegin();
+        auto y = v.begin();
 
         for ( ;; ) {
-            if ( x == safeEnd() )
+            if ( x == end() )
                 return std::make_tuple(false, i);
 
             if ( *x++ != *y++ )
                 break;
 
-            if ( y == v.safeEnd() )
+            if ( y == v.end() )
                 return std::make_tuple(true, i);
         }
     }

--- a/hilti/src/rt/types/regexp.cc
+++ b/hilti/src/rt/types/regexp.cc
@@ -228,7 +228,7 @@ int32_t RegExp::find(const Bytes& data) const {
 }
 
 static Bytes _subslice(const Bytes& data, jrx_offset so, jrx_offset eo) {
-    return Bytes(data.sub(data.safeBegin() + so, data.safeBegin() + eo));
+    return Bytes(data.sub(data.begin() + so, data.begin() + eo));
 }
 
 std::tuple<int32_t, Bytes> RegExp::findSpan(const Bytes& data) const {

--- a/hilti/src/rt/types/stream.cc
+++ b/hilti/src/rt/types/stream.cc
@@ -109,7 +109,7 @@ std::tuple<bool, SafeConstIterator> View::find(const Bytes& v, const SafeConstIt
     if ( v.isEmpty() )
         return std::make_tuple(true, n ? n : _begin);
 
-    auto first = *v.safeBegin();
+    auto first = *v.begin();
 
     for ( auto i = UnsafeConstIterator(n ? n : _begin); true; ++i ) {
         if ( i == UnsafeConstIterator(safeEnd()) )

--- a/zeek/plugin/include/runtime-support.h
+++ b/zeek/plugin/include/runtime-support.h
@@ -290,7 +290,7 @@ inline Val* to_val(const hilti::rt::Bytes& b, BroType* target, std::string_view 
     if ( target->Tag() != ::TYPE_STRING )
         throw TypeMismatch("string", target, location);
 
-    return new ::StringVal(b);
+    return new ::StringVal(b.str());
 }
 
 /**

